cocoa/interaction.m

```diff
@@ -0,0 +1,890 @@
+    NSAttributedString *a = [[NSAttributedString alloc] initWithString:s attributes:nil];
+
+    free(buf);
+    [s release];
+    return [a autorelease];
+}
+
+- (NSUInteger)characterIndexForPoint:(NSPoint)aPoint {
+    NSLog(@"WARNING: unimplemented characterIndexForPoint:%@", NSStringFromPoint(aPoint));
+    NSLog(@"If you see this message, file a bug about \"characterIndexForPoint:\" !");
+    return 0;
+}
+
+- (NSRect)firstRectForCharacterRange:(NSRange)aRange actualRange:(NSRangePointer)actualRange {
+    CGRect loc = find_ui_object_in_window(&edit_get_active()->panel);
+    //NSLog(@"%@", NSStringFromRect(loc));
+    return [self.window convertRectToScreen:loc];
+}
+
+@end
+
+// FIXME: asda
+static char_t clip_data[65536];
+
+void setselection(char_t *data, STRING_IDX length) {
+}
+
+int getbuf(char_t *ptr, size_t len, int value) {
+    int ret = 0;
+    if (edit_active()) {
+        // FIXME: asfasg
+        ret = edit_copy(ptr, len);
+    } else if(sitem->item == ITEM_FRIEND) {
+        ret = messages_selection(&messages_friend, ptr, len, value);
+    } else {
+        ret = messages_selection(&messages_group, ptr, len, value);
+    }
+    
+    return ret;
+}
+
+void copy(int value) {
+    int len;
+    NSString *strtocopy;
+
+    len = getbuf(clip_data, sizeof(clip_data), value);
+
+    strtocopy = [[NSString alloc] initWithBytes:clip_data length:len encoding:NSUTF8StringEncoding];
+    if(len) {
+        [[NSPasteboard generalPasteboard] clearContents];
+        [[NSPasteboard generalPasteboard] writeObjects:@[strtocopy]];
+    }
+    [strtocopy release];
+}
+
+void paste(void) {
+    NSPasteboard *pb = [NSPasteboard generalPasteboard];
+    NSArray *arr = [pb readObjectsForClasses:@[NSImage.class, NSString.class] options:nil];
+    id string_or_img = arr.firstObject;
+
+    if ([string_or_img isKindOfClass:NSString.class]) {
+        NSString *str = string_or_img;
+        if (edit_active()) {
+            edit_paste((char_t *)str.UTF8String, [str lengthOfBytesUsingEncoding:NSUTF8StringEncoding], 0);
+        } else {
+            NSBeep();
+        }
+    } else /* NSImage */ {
+        [string_or_img lockFocus];
+        NSBitmapImageRep *bmp = [[NSBitmapImageRep alloc] initWithFocusedViewRect:(CGRect){CGPointZero, [string_or_img size]}];
+        [string_or_img unlockFocus];
+
+        CGImageRef img = CGImageRetain(bmp.CGImage);
+        UTOX_NATIVE_IMAGE *i = malloc(sizeof(UTOX_NATIVE_IMAGE));
+        i->scale = 1.0;
+        i->image = img;
+
+        CFMutableDataRef dat = CFDataCreateMutable(kCFAllocatorDefault, 0);
+        CGImageDestinationRef dest = CGImageDestinationCreateWithData(dat, kUTTypePNG, 1, NULL);
+        CGImageDestinationAddImage(dest, img, NULL);
+        CGImageDestinationFinalize(dest);
+        CFRelease(dest);
+
+        size_t size = CFDataGetLength(dat);
+        uint8_t *owned_ptr = malloc(size);
+        memcpy(owned_ptr, CFDataGetBytePtr(dat), size);
```

Comment:
if NULL undefined behaviour... wanna wrap this with an error just incase malloc fails?